isisd: add vrf support for redistribution function in ISIS

aldobrrrr opened the pull request:
Now it's possible to redistribute from another protocol routes which are
inside a specific vrf.

Example of possible configuration:

```
!
vrf isis-l1
 ipv6 route fd00::/60 blackhole
 exit-vrf
!
interface one vrf isis-l1
 ipv6 router isis COMMON vrf isis-l1
 isis circuit-type level-1
!
router isis COMMON vrf isis-l1
 is-type level-1
 net fd.0000.0000.0000.0001.00
 redistribute ipv6 static level-1
 topology ipv6-unicast
!
```

Signed-off-by: Emanuele Altomare <emanuele@common-net.org>

idryzhov commented:
Looks like a correct idea, but I have the following question:
If we configure this when the VRF is not yet created in the kernel, then `VRF_UNKNOWN` will be passed to `zclient_redistribute`. Later, when the VRF is created, shouldn't we update this?

P.S. Same issue in OSPF as I see.

idryzhov reviewed (changes requested):
Looks like a correct idea, but I have the following question:
If we configure this when the VRF is not yet created in the kernel, then VRF_UNKNOWN will be passed to zclient_redistribute. Later, when the VRF is created, shouldn't we update this?

P.S. Same issue in OSPF as I see.

pguibert6WIND commented:
> Looks like a correct idea, but I have the following question:
> If we configure this when the VRF is not yet created in the kernel, then VRF_UNKNOWN will be passed to zclient_redistribute. Later, when the VRF is created, shouldn't we update this?
> 
> P.S. Same issue in OSPF as I see.

I think this should be processed as a separate issue.
only BGP is known to work without zebra.
the other daemons are are zebra dependent.

idryzhov commented:
@pguibert6WIND I don't understand your comment.

After this PR, if you enter the config:
```
router isis test vrf red
 redistribute something
```
when the vrf red doesn't exist in the kernel then `zclient_redistribute` will be called with VRF_UNKNOWN vrf id.
Isn't this incorrect? This id won't be even updated when we later actually create the vrf.

polychaeta reviewed (dismissed):
Thanks for your contribution to FRR!

* `sprintf, strcat, strcpy, inet_ntoa` are banned; please use `snprintf, strlcat, strlcpy, inet_ntop`

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

aldobrrrr commented:
CI:rerun

aldobrrrr commented:
> Looks like a correct idea, but I have the following question:
> If we configure this when the VRF is not yet created in the kernel, then VRF_UNKNOWN will be passed to zclient_redistribute. Later, when the VRF is created, shouldn't we update this?
> 
> P.S. Same issue in OSPF as I see.

Hi, I've taken your point and I've added the code which fixes this behaviour. Now when a VRF is added/deleted the subscriptions with Zebra are also updated. The only problem is that the test at tests/topotests/isis-topo1-vrf fails.
When I run this test, the topology is the same like it should, but some routes have different metrics. I didn't touch metric in any way as you can see in my modifications, it is a side effect which could be not necessarily wrong, but I need help to understand.

I tried my code with the help of debugger and in every case the code behaves like it should. If there is an error I don't see it.

Can you help me?

Thank you.

P.S. The parsing function for "show isis vrf ... topology ..." inside the test code is buggy and should be fixed: it doesn't handle correctly spaces (and some vertex type identifiers have space), so for some routes in "expected topology" JSON file, you will find strange things like metric in nexthop field.

idryzhov commented:
@aldobrrrr please try to rebase on top of the latest master. There were some fixes for CI failures.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/2130ef312d26dc38fe15c2cd35213b24/raw/0d3c29a37ddca0735a7dfd222573494d156e3bd7/cr_8215_1617014528.diff | git apply
```

</p>

<p>

```diff
diff --git a/isisd/isis_redist.c b/isisd/isis_redist.c
index 11f547685..632c64e0f 100644
--- a/isisd/isis_redist.c
+++ b/isisd/isis_redist.c
@@ -222,8 +222,8 @@ static void isis_redist_ensure_default(struct isis *isis, int family)
 
 static void delete_ext_reach_routes(struct isis_redist *redist)
 {
-	struct route_table *er_table = get_ext_reach(redist->area,
-						redist->family, redist->level);
+	struct route_table *er_table =
+		get_ext_reach(redist->area, redist->family, redist->level);
 	struct isis_ext_info *info;
 	struct route_node *rn;
 
@@ -446,10 +446,10 @@ static void isis_redist_update_zebra_subscriptions(struct isis *isis)
 
 			if (do_subscribe[protocol][type])
 				isis_zebra_redistribute_set(afi, type,
-						isis->vrf_id);
+							    isis->vrf_id);
 			else
 				isis_zebra_redistribute_unset(afi, type,
-						isis->vrf_id);
+							      isis->vrf_id);
 		}
 }
 
@@ -486,11 +486,11 @@ void isis_redist_set(struct isis_area *area, int level, int family, int type,
 	redist->level = level;
 
 	memset(redist->vrf_name, 0, sizeof(redist->vrf_name));
-	if (isis->name && strlen(isis->name) &&
-	    strlen(isis->name) < sizeof(redist->vrf_name) &&
-	    strcmp(isis->name, vrf_get_default_name())) {
+	if (isis->name && strlen(isis->name)
+	    && strlen(isis->name) < sizeof(redist->vrf_name)
+	    && strcmp(isis->name, vrf_get_default_name())) {
 		snprintf(redist->vrf_name, sizeof(redist->vrf_name), "%s",
-			isis->name);
+			 isis->name);
 
 		hook_register_arg(isis_vrf_enable_hook, isis_redist_vrf_enable,
 				  redist);
diff --git a/isisd/isis_zebra.c b/isisd/isis_zebra.c
index 9e9185096..e531eb4eb 100644
--- a/isisd/isis_zebra.c
+++ b/isisd/isis_zebra.c
@@ -528,8 +528,8 @@ void isis_zebra_redistribute_set(afi_t afi, int type, vrf_id_t vrf_id)
 		zclient_redistribute_default(ZEBRA_REDISTRIBUTE_DEFAULT_ADD,
 					     zclient, afi, vrf_id);
 	else
-		zclient_redistribute(ZEBRA_REDISTRIBUTE_ADD, zclient, afi,
-				     type, 0, vrf_id);
+		zclient_redistribute(ZEBRA_REDISTRIBUTE_ADD, zclient, afi, type,
+				     0, vrf_id);
 }
 
 void isis_zebra_redistribute_unset(afi_t afi, int type, vrf_id_t vrf_id)
diff --git a/isisd/isisd.c b/isisd/isisd.c
index ef34cdd58..bb0f0812c 100644
--- a/isisd/isisd.c
+++ b/isisd/isisd.c
@@ -91,8 +91,8 @@ DEFINE_MTYPE(ISISD, ISIS_ACL_NAME,    "ISIS access-list name");
 
 DEFINE_QOBJ_TYPE(isis_area);
 
-DEFINE_HOOK(isis_vrf_enable_hook, (struct vrf *vrf), (vrf))
-DEFINE_HOOK(isis_vrf_disable_hook, (struct vrf *vrf), (vrf))
+DEFINE_HOOK(isis_vrf_enable_hook, (struct vrf * vrf), (vrf))
+DEFINE_HOOK(isis_vrf_disable_hook, (struct vrf * vrf), (vrf))
 
 /* ISIS process wide configuration. */
 static struct isis_master isis_master;
diff --git a/isisd/isisd.h b/isisd/isisd.h
index 425ead4ba..1baeb2079 100644
--- a/isisd/isisd.h
+++ b/isisd/isisd.h
@@ -67,8 +67,8 @@ extern void isis_cli_init(void);
 
 #define SNMP_CIRCUITS_MAX (512)
 
-DECLARE_HOOK(isis_vrf_enable_hook, (struct vrf *vrf), (vrf))
-DECLARE_HOOK(isis_vrf_disable_hook, (struct vrf *vrf), (vrf))
+DECLARE_HOOK(isis_vrf_enable_hook, (struct vrf * vrf), (vrf))
+DECLARE_HOOK(isis_vrf_disable_hook, (struct vrf * vrf), (vrf))
 
 extern struct zebra_privs_t isisd_privs;
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

aldobrrrr commented:
CI:rerun

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/f3ceb1c430ccef2f3e185d7b748cdfdb/raw/945a74d16eac8fd689e879ab20067117c9b949c0/cr_8215_1617035199.diff | git apply
```

</p>

<p>

```diff
diff --git a/isisd/isis_redist.c b/isisd/isis_redist.c
index 11f547685..632c64e0f 100644
--- a/isisd/isis_redist.c
+++ b/isisd/isis_redist.c
@@ -222,8 +222,8 @@ static void isis_redist_ensure_default(struct isis *isis, int family)
 
 static void delete_ext_reach_routes(struct isis_redist *redist)
 {
-	struct route_table *er_table = get_ext_reach(redist->area,
-						redist->family, redist->level);
+	struct route_table *er_table =
+		get_ext_reach(redist->area, redist->family, redist->level);
 	struct isis_ext_info *info;
 	struct route_node *rn;
 
@@ -446,10 +446,10 @@ static void isis_redist_update_zebra_subscriptions(struct isis *isis)
 
 			if (do_subscribe[protocol][type])
 				isis_zebra_redistribute_set(afi, type,
-						isis->vrf_id);
+							    isis->vrf_id);
 			else
 				isis_zebra_redistribute_unset(afi, type,
-						isis->vrf_id);
+							      isis->vrf_id);
 		}
 }
 
@@ -486,11 +486,11 @@ void isis_redist_set(struct isis_area *area, int level, int family, int type,
 	redist->level = level;
 
 	memset(redist->vrf_name, 0, sizeof(redist->vrf_name));
-	if (isis->name && strlen(isis->name) &&
-	    strlen(isis->name) < sizeof(redist->vrf_name) &&
-	    strcmp(isis->name, vrf_get_default_name())) {
+	if (isis->name && strlen(isis->name)
+	    && strlen(isis->name) < sizeof(redist->vrf_name)
+	    && strcmp(isis->name, vrf_get_default_name())) {
 		snprintf(redist->vrf_name, sizeof(redist->vrf_name), "%s",
-			isis->name);
+			 isis->name);
 
 		hook_register_arg(isis_vrf_enable_hook, isis_redist_vrf_enable,
 				  redist);
diff --git a/isisd/isis_zebra.c b/isisd/isis_zebra.c
index 9e9185096..e531eb4eb 100644
--- a/isisd/isis_zebra.c
+++ b/isisd/isis_zebra.c
@@ -528,8 +528,8 @@ void isis_zebra_redistribute_set(afi_t afi, int type, vrf_id_t vrf_id)
 		zclient_redistribute_default(ZEBRA_REDISTRIBUTE_DEFAULT_ADD,
 					     zclient, afi, vrf_id);
 	else
-		zclient_redistribute(ZEBRA_REDISTRIBUTE_ADD, zclient, afi,
-				     type, 0, vrf_id);
+		zclient_redistribute(ZEBRA_REDISTRIBUTE_ADD, zclient, afi, type,
+				     0, vrf_id);
 }
 
 void isis_zebra_redistribute_unset(afi_t afi, int type, vrf_id_t vrf_id)
diff --git a/isisd/isisd.c b/isisd/isisd.c
index d3bf170a5..945774923 100644
--- a/isisd/isisd.c
+++ b/isisd/isisd.c
@@ -91,8 +91,8 @@ DEFINE_MTYPE(ISISD, ISIS_ACL_NAME,    "ISIS access-list name");
 
 DEFINE_QOBJ_TYPE(isis_area);
 
-DEFINE_HOOK(isis_vrf_enable_hook, (struct vrf *vrf), (vrf));
-DEFINE_HOOK(isis_vrf_disable_hook, (struct vrf *vrf), (vrf));
+DEFINE_HOOK(isis_vrf_enable_hook, (struct vrf * vrf), (vrf));
+DEFINE_HOOK(isis_vrf_disable_hook, (struct vrf * vrf), (vrf));
 
 /* ISIS process wide configuration. */
 static struct isis_master isis_master;
diff --git a/isisd/isisd.h b/isisd/isisd.h
index ec0257b9f..e2a84392a 100644
--- a/isisd/isisd.h
+++ b/isisd/isisd.h
@@ -67,8 +67,8 @@ extern void isis_cli_init(void);
 
 #define SNMP_CIRCUITS_MAX (512)
 
-DECLARE_HOOK(isis_vrf_enable_hook, (struct vrf *vrf), (vrf));
-DECLARE_HOOK(isis_vrf_disable_hook, (struct vrf *vrf), (vrf));
+DECLARE_HOOK(isis_vrf_enable_hook, (struct vrf * vrf), (vrf));
+DECLARE_HOOK(isis_vrf_disable_hook, (struct vrf * vrf), (vrf));
 
 extern struct zebra_privs_t isisd_privs;
 

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

aldobrrrr commented:
> @aldobrrrr please try to rebase on top of the latest master. There were some fixes for CI failures.

Thank you for your reply, I've done the rebase, but nothing has changed. The problem is, like I wrote in my previous comment, that the "tests/topotests/isis-topo1-vrf" test fails due to the expected metrics. With my patch all the metrics for distributed routes have a constant decrease of 10. I haven't touched metrics so it's a side effect of my modifications.

I would need some help to understand if the behaviour is right and only the test should be updated or it's wrong and some other piece of code need to be patched.

vjardin commented:
LGTM. Beside the marginal topotest issue, this feature is a must have for isisd.

For the time being it needs to be rebased.

idryzhov commented:
@vjardin the issue is already fixed by https://github.com/FRRouting/frr/pull/8628. I forgot to close this PR, it is not needed anymore.

aldobrrrr commented:
> @vjardin the issue is already fixed by #8628. I forgot to close this PR, it is not needed anymore.

Already is not correct, because my PR comes first and it is "not needed anymore" because you copied my PR, rewriting it in a "easier way" (as you wrote in your PR, but is only an uglier way), without solve or understand the problem relative to the failing test, but simply patching the test itself.

So you have decided unilaterally that my time isn't important and, maybe worse, that in our profession the understanding of things is not so important. You are a very poor developer but more important you are a very poor man because I asked for help to understand the problem under the metrics change and you have done what you done.

idryzhov commented:
@aldobrrrr, let's be crystal clear here.

You proposed a PR that was not reviewed and accepted by the community.
I proposed another PR, which solved several problems in ISIS, one of them being the one you were trying to fix in your PR. I did this using an absolutely different approach **without using any of your code** – the only common part is passing `vrf_id` to `zclient_redistribute`, which is unavoidable. Note that in my PR, I mentioned yours right away, saying that mine is another approach for the same issue, not trying to take credit for finding the problem.
So don't you think that calling me a poor developer and a poor man is too much for the totally legit work?

Thanks for calling my code ugly, by the way. I suppose it means that I, the reviewers and developers who used the same approach in other daemons, are all stupid people, right?

qlyoung commented:
@aldobrrrr I understand why you're upset but you need to be professional. We are an open source community, these conflicts inevitably happen from time to time, and when they do it's the responsibility of all parties involved to work towards a resolution as gracefully as possible. I see where you're coming from but please don't assume bad faith.

@idryzhov in the future please try to communicate with someone who's work you're obviating. You did mention this PR in yours, but a @mention might have helped us avoid this situation (I'm assuming that you didn't speak to @aldobrrrr through some other channel about this PR; if you did please clarify). Also, closing a PR when you did the obviating work leaves room for interpretations like the above - a better choice would have been to poke another maintainer to resolve this one.

To both of you I request that we tone down the heat on this conversation.

---

So, at this stage we have @idryzhov's work merged. @aldobrrrr, aside from the issues mentioned above, what's your technical criticism of his solution? What would you do better? I understand that it would have been preferable to have more communication in advance but we now have to deal with things as they are. If you have technical criticisms let's hear them and get them addressed.

idryzhov commented:
Thank you for helping with this, Quentin. I'll do like you suggest if something like this happens again.

Another thing we can do is provide some guidance in the developer docs on how to deal with such problems. Something similar to [Kernel's doc](https://www.kernel.org/doc/html/latest/process/submitting-patches.html#using-reported-by-tested-by-reviewed-by-suggested-by-and-fixes). "Reported-by" or "Suggested-by" tag would be appropriate here, I suppose.

aldobrrrr commented:
> @aldobrrrr I understand why you're upset but you need to be professional. We are an open source community, these conflicts inevitably happen from time to time, and when they do it's the responsibility of all parties involved to work towards a resolution as gracefully as possible. I see where you're coming from but please don't assume bad faith.
> 
> @idryzhov in the future please try to communicate with someone who's work you're obviating. You did mention this PR in yours, but a @mention might have helped us avoid this situation (I'm assuming that you didn't speak to @aldobrrrr through some other channel about this PR; if you did please clarify). Also, closing a PR when you did the obviating work leaves room for interpretations like the above - a better choice would have been to poke another maintainer to resolve this one.
> 
> To both of you I request that we tone down the heat on this conversation.
> 
> So, at this stage we have @idryzhov's work merged. @aldobrrrr, aside from the issues mentioned above, what's your technical criticism of his solution? What would you do better? I understand that it would have been preferable to have more communication in advance but we now have to deal with things as they are. If you have technical criticisms let's hear them and get them addressed.

Thank you for your reply @qlyoung, I report some considerations about the implementations as you asked me.

Considerations about my code:

1) It adds the area, family, type and level to the redist struct, which are keys for redist struct inside redist_settings multidimensional array. This can be usefull to take some decisions without using nested iterations to access the array also when only one redist struct is needed. (the function isis_redist_update_subscriptions() is an example of what I say with nested iterations, and the fields added into the struct can help, in the future, to reduce the use of that function).

2) It adds in redist struct the name of VRF so we can bind to the name which comes from configuration and not to the ID which can change. The kernel, as far as I know, only guarantees a monotonic ID, but when we disable and enable again a VRF with the same name the ID will most probably change.

3) It adds 2 hooks "isis_vrf_enable_hook" and "isis_vrf_disable_hook": in this way it's possible to decouple the functions isisd_vrf_enable and isisd_vrf_disable in isisd.c from the precessing of isis_redist.c and it's possible to register only the redist structs which are enabled without executing nested iterations over all redist_settings multidimensional array.

Considerations about @idryzhov  code:

1) Why copy and paste the ugly static (so not intended by the author to be exposed) isis_redist_update_subscriptions() in another file (isis_set_redist_vrf_bitmaps(struct isis *isis, bool set) in isisd.c) when we can use hooks to decouple and we can call directly isis_zebra_redistribute_set() which set the bitmap for us, all inside the isis_redist.c?

2) Why does isisd.c need to access a data structure of isis_redist.c? isis_set_redist_vrf_bitmaps sounds like a behavior proper of isis_redist.c, doesn't it? I think concepts like encapsulation and single responsibility are very important, especially in big projects like FRR.

qlyoung commented:
I don't have any strong opinions about encapsulation, single responsibility or other aspects of program structuring in this particular case, as this daemon is largely outside of my expertise.

Regarding this:
> It adds in redist struct the name of VRF so we can bind to the name which comes from configuration and not to the ID which can change. The kernel, as far as I know, only guarantees a monotonic ID, but when we disable and enable again a VRF with the same name the ID will most probably change.

Can you elaborate what you mean by "bind to the name"? I assume you don't mean an actual `setsockopt` type bind, you mean we're saving the name aside here for reference within isisd later for some other use case right?

@idryzhov can you have a look at his comments and advise what your thoughts are?

aldobrrrr commented:
> Regarding this:
> 
> > It adds in redist struct the name of VRF so we can bind to the name which comes from configuration and not to the ID which can change. The kernel, as far as I know, only guarantees a monotonic ID, but when we disable and enable again a VRF with the same name the ID will most probably change.
> 
> Can you elaborate what you mean by "bind to the name"? I assume you don't mean an actual `setsockopt` type bind, you mean we're saving the name aside here for reference within isisd later for some other use case right?

Yes, sorry I don't express myself very well in english. I mean the use of the VRF name as a reference. I think it's more stable the name instead of the ID for the reason explained in my previous message.

qlyoung commented:
So you're saying that by saving the VRF ID instead of the name, if the VRF device is destroyed and then recreated, that will cause problems in the current code? Won't we receive the VRF destroy notification, delete the existing structures that reference that VRF, and then recreate them when the VRF create notification comes in? There should be no concept of preserving a vrf between a destroy and create call except where configuration references it - and here it must be stored as a string anyway, since that's all the CLI accepts.

aldobrrrr commented:
> So you're saying that by saving the VRF ID instead of the name, if the VRF device is destroyed and then recreated, that will cause problems in the current code? Won't we receive the VRF destroy notification, delete the existing structures that reference that VRF, and then recreate them when the VRF create notification comes in? There should be no concept of preserving a vrf between a destroy and create call except where configuration references it - and here it must be stored as a string anyway, since that's all the CLI accepts.

No, I'm saying that the code saves in the redist struct, which is the memory transposition of redistribution configuration, also the vrf name which is part of configuration. In fact the vrf name is taken by the configuration of the isis instance (`router isis COMMON vrf isis-l1`). This information is implicit now but no one stops us, in the future, to add this parameter explicitly in the configuration to allow the redistribute between vrf. For example with

`router isis COMMON vrf isis-l1
 is-type level-1
 net fd.0000.0000.0000.0001.00
 redistribute ipv6 kernel level-1 vrf example-vrf 
 topology ipv6-unicast`

we could redistribute from the kernel (vrf "example-vrf") into the ISIS level-1 (in the "isis-l1" vrf).

In every case the vrf name is a redistribution parameter and I made it explicit.

aldobrrrr commented:
In general I want to emphasize that in my code with the callbacks we can register only the redistributions really active to receive the updates from the underlying kernel situation relative to the vrf. It is very unlikely that, for an ISIS instance, we have the redistributions from all (or close to all) available protocols active. So iterate over all the multidimensional array of redist structs to check if any is active is not a good idea. The code turns rapidly in a mess of nested iterations and conditions.
